broadinstitute/single_cell_portal_core
# Allow re-upload of same study file after successful deletion (SCP-3451)

A regression was introduced during the Rails 6.1  & `CarrierWave` [upgrade](#1030) where the `StudyFile` attribute `upload_file_name` is not always being successfully cleared when invoking `DeleteQueueJob`, which results in the user not being able to upload a file with the same filename after deleting it.  This is a common workflow when addressing file validation issues.  The error is not completely reproducible - it seems somewhat random as to when it happens - although files uploaded prior to the upgrade seem much more likely to encounter this issue.

This update fixes the problem via the following mechanisms:
1. `queued_for_deletion: false` is now appended to queries where SCP is trying to find a file in a study using `upload_file_name`
2. Validating uniqueness of `upload_file_name` ignores files where `queued_for_deletion: true`
3. `DeleteQueueJob` now automatically deletes `StudyFile` documents once all parsed data has been removed from the database

The third change is the main mechanism for correcting the regression, whereas the first & second changes address potential race conditions while `DeleteQueueJob` runs.

MANUAL TESTING

1. Sign in and create a new study (or use an existing study)
2. Add a `Gene List` file, using `test/test_data/marker_1_gene_list.txt`
3. Add a doc/miscellaneous file, using `test/test_data/mock_study_doc_upload.txt`
4. Delete both files
5. Without refreshing the page, re-upload both files and note that neither throws an error saying the file already exists

This PR satisfies SCP-3451.
